x-pack/plugins/security_solution/server/endpoint/services/actions/create/write_action_to_indices.ts

```diff
@@ -0,0 +1,220 @@
+    const fleetActionDoc = {
+      ...doc.EndpointActions,
+      '@timestamp': doc['@timestamp'],
+      agents,
+      timeout: 300, // 5 minutes
+      user_id: doc.user.id,
+    };
+    const fleetActionDocSignature = await endpointContext.service
+      .getMessageSigningService()
+      .sign(fleetActionDoc);
+    const signedFleetActionDoc = {
+      ...fleetActionDoc,
+      signed: {
+        data: fleetActionDocSignature.data.toString('base64'),
+        signature: fleetActionDocSignature.signature,
+      },
+    };
+    // write actions to .fleet-actions index
+    try {
+      const fleetActionIndexResult = await esClient.index<EndpointAction>(
+        {
+          index: AGENT_ACTIONS_INDEX,
+          body: signedFleetActionDoc,
+          refresh: 'wait_for',
+        },
+        {
+          meta: true,
+        }
+      );
+
+      if (fleetActionIndexResult.statusCode !== 201) {
+        throw new Error(fleetActionIndexResult.body.result);
+      }
+    } catch (e) {
+      // create entry in .logs-endpoint.action.responses-default data stream
+      // when writing to .fleet-actions fails
+      if (doesLogsEndpointActionsDsExist) {
+        await createFailedActionResponseEntry({
+          esClient,
+          doc: {
+            '@timestamp': moment().toISOString(),
+            agent: doc.agent,
+            EndpointActions: {
+              action_id: doc.EndpointActions.action_id,
+              completed_at: moment().toISOString(),
+              started_at: moment().toISOString(),
+              data: doc.EndpointActions.data,
+            },
+          },
+          logger,
+        });
+      }
+
+      throw e;
+    }
+  }
+};
+
+const createFailedActionResponseEntry = async ({
+  esClient,
+  doc,
+  logger,
+}: {
+  esClient: ElasticsearchClient;
+  doc: LogsEndpointActionResponse;
+  logger: Logger;
+}): Promise<void> => {
+  try {
+    await esClient.index<LogsEndpointActionResponse>({
+      index: `${ENDPOINT_ACTION_RESPONSES_DS}-default`,
+      body: {
+        ...doc,
+        error: {
+          code: failedFleetActionErrorCode,
+          message: 'Failed to deliver action request to fleet',
+        },
+      },
+    });
+  } catch (error) {
+    logger.error(wrapErrorIfNeeded(error));
+  }
+};
+
+const addRuleInfoToAction = (
+  payload: CreateActionPayload
+):
+  | {
+      rule?: { id: string; name: string };
+    }
+  | undefined => {
+  if (payload.rule_id && payload.rule_name) {
+    return { rule: { id: payload.rule_id, name: payload.rule_name } };
+  }
+};
+
+const getActionParameters = (
+  action: CreateActionPayload
+): ResponseActionsExecuteParameters | Readonly<{}> | undefined => {
+  // set timeout to 4h (if not specified or when timeout is specified as 0) when command is `execute`
+  if (action.command === 'execute') {
+    const actionRequestParams = action.parameters as ResponseActionsExecuteParameters;
+    if (typeof actionRequestParams?.timeout === 'undefined') {
```

Comment:
I usually prefer checking the type of the value when checking for undefined cause in case `timeout` is not declared I won't get an error. See [here](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/undefined#examples)